### PR TITLE
[iOS] Fix the mess we had with the logs.

### PR DESCRIPTION
**Pull request by mandel-macaque**
The initial implementation of the logs, yet useful, it was wrong. It
made it very easy to pass Logs that do have certain missing properties
and methods to other classes that used those methods, resulting in
runtime issues.

We move to have different interfaces that will allow to say which is the
min amount of methods to be implemented (ILog) or the max one
(IFileBackedLog) so that the compiler will help us to not pass a
MemoryLog or a ConsoleLog to those methods that will like to use, for
example, the FullPath of the log.

This might look like a breaking change for xamarin-macios, but whatever
it breaks was a hidden bug.

### CODE REVIEW

File: src/Microsoft.DotNet.XHarness.Common/Logging/ILog.cs [hunk 1]
```diff
@@ -13,15 +13,29 @@ public interface ILog : IDisposable
...
         void Flush();
     }
+
+    /// <summary>
+    /// Interface to be implemented by those loggers that provide a StreamReader to read
+    /// the already added info.
+    /// </summary>
+    public interface IReadable
+    {
+        StreamReader GetReader();
```
Comment (premun) [inline src/Microsoft.DotNet.XHarness.Common/Logging/ILog.cs:31]: Why do we have this `IReadable`? Why not put `GetReader()` straight to `IReadableLog`?

File: src/Microsoft.DotNet.XHarness.iOS.Shared/Logging/ILogs.cs [hunk 2]
```diff
@@ -8,22 +8,22 @@
 
 namespace Microsoft.DotNet.XHarness.iOS.Shared.Logging
 {
-    public interface ILogs : IList<ILog>, IDisposable
+    public interface ILogs : IList<IFileBackedLog>, IDisposable
```
Comment (premun) [inline src/Microsoft.DotNet.XHarness.iOS.Shared/Logging/ILogs.cs:11]: I guess these are then `IFileBackedLogs`?

Comment (premun) [reply src/Microsoft.DotNet.XHarness.iOS.Shared/Logging/ILogs.cs:11]: Also wondering - why not just `IFileLog`?

File: src/Microsoft.DotNet.XHarness.iOS.Shared/Logging/ReadableLog.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+using System.IO;
+using Microsoft.DotNet.XHarness.Common.Logging;
+
+namespace Microsoft.DotNet.XHarness.iOS.Shared.Logging
+{
+    public abstract class ReadableLog : Log, IReadableLog
+    {
+
+        protected ReadableLog(string description = null) : base(description) { }
```
Comment (premun) [inline src/Microsoft.DotNet.XHarness.iOS.Shared/Logging/ReadableLog.cs:9]:
```suggestion
        protected ReadableLog(string? description = null) : base(description) { }
```

File: src/Microsoft.DotNet.XHarness.iOS.Shared/Logging/ReadableLog.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+        public abstract StreamReader GetReader();
+    }
+
+    public abstract class FileBackedLog : ReadableLog, IFileBackedLog
+    {
+        protected FileBackedLog(string description = null) : base(description) { }
```
Comment (premun) [inline src/Microsoft.DotNet.XHarness.iOS.Shared/Logging/ReadableLog.cs:16]:
```suggestion
        protected FileBackedLog(string? description = null) : base(description) { }
```

File: src/Microsoft.DotNet.XHarness.iOS.Shared/TestReporter.cs [hunk 5]
```diff
@@ -30,14 +30,14 @@ namespace Microsoft.DotNet.XHarness.iOS.Shared
     public class TestReporter : ITestReporter
     {
 
-        const string timeoutMessage = "Test run timed out after {0} minute(s).";
-        const string completionMessage = "Test run completed";
-        const string failureMessage = "Test run failed";
+        const string _timeoutMessage = "Test run timed out after {0} minute(s).";
+        const string _completionMessage = "Test run completed";
+        const string _failureMessage = "Test run failed";
```
Comment (premun) [inline src/Microsoft.DotNet.XHarness.iOS.Shared/TestReporter.cs:35]:
```suggestion
        private const string _timeoutMessage = "Test run timed out after {0} minute(s).";
        private const string _completionMessage = "Test run completed";
        private const string _failureMessage = "Test run failed";
```

and below

Comment (premun) [reply src/Microsoft.DotNet.XHarness.iOS.Shared/TestReporter.cs:35]: Also I think ConstsShouldBeLikeThis

File: src/Microsoft.DotNet.XHarness.Common/Logging/ILog.cs (hunk 1 above)
Comment (mandel-macaque) [reply src/Microsoft.DotNet.XHarness.Common/Logging/ILog.cs:31]: Seemed to be a good idea at the time, but know that you mention it, I'll remove that, lets see what happens, no the compiler helps :)

File: src/Microsoft.DotNet.XHarness.iOS.Shared/Logging/ILogs.cs (hunk 2 above)
Comment (mandel-macaque) [reply src/Microsoft.DotNet.XHarness.iOS.Shared/Logging/ILogs.cs:11]: I tried, less freedom :/

File: src/Microsoft.DotNet.XHarness.Common/Logging/ILog.cs (hunk 1 above)
Comment (mandel-macaque) [reply src/Microsoft.DotNet.XHarness.Common/Logging/ILog.cs:27]: There are classes that just care about this interface. Gives us freedom.

**Pull request merged**